pandas/core/generic.py

```diff
@@ -5257,7 +5257,12 @@ def _reindex_with_indexers(
             # If we've made a copy once, no need to make another one
             copy = False
 
-        if (copy or copy is None) and new_data is self._mgr:
+        if (
+            (copy or copy is None)
+            and new_data is self._mgr
+            or not copy
+            and using_copy_on_write()
+        ):
```

Comment:
Can you explain a bit how this change is related? We are passing here a Manager object to the constructor, I suppose, and you only changed the constructor when it receives a DataFrame?

Reply:
Ah sorry, this is unrelated to the constructor change, I'll make a separate pr. 

Currently, when reindexing with copy=False we don't track references, causing problems.

Reply:
I also have some skips in https://github.com/pandas-dev/pandas/pull/50536 related to `reindex`, there are several things not yet fully working